### DISCUSSION TITLE
Building natively for Windows

### DISCUSSION THREAD

**Discussion by Anton-V-K** [colobot/colobot#1535] [General]
What is the current way to build the application natively for Windows, for instance, using msbuild from VisualStudio (VS2019)?
The [official instruction](https://github.com/colobot/colobot/blob/master/INSTALL.md) states it is possible and references the [old forum's post](https://colobot.info/forum/showthread.php?tid=595&pid=5831#pid5831) (relevant only for VS2013?), which relies on attached pre-built dependencies (none of them is available nowadays).
I tried to set up own environment (for `dev` branch) from scratch feeding required modules to `cmake`. It was quite easy with SDL2-related libraries (you can just download them in a pre-build form and expose their locations with environment variables), but `libpng` is another story - `cmake` can't find it along with `zlib`.
Is anyone building for Windows natively (i.e. not using MXE)? Any hints how to set up the environment?
BTW according to #1306 the build system is being redesigned, so probably this isssue will also be addressed.

**Comment by tomaszkax86** [comment]
There's a new set of instructions on how to build Colobot natively on Windows: https://github.com/colobot/colobot/wiki/How-to-Build-Colobot:-Gold-Edition-Using-MSVC. We probably just need to cleanup Wiki because some of the stuff it has is outdated.

**Reply by Anton-V-K** [reply]
Thanks for the link!
I've followed the new instruction and succeeded in setting up the environment (`vcpkg` makes it all a lot easier!).
VS2019 processes 435 files (sources and translations), but unfortunately ends up with the error:
```
...\colobot\out\build\x64-Debug\LINK : fatal error LNK1104: cannot open file 'src\CBot\CBot.lib'
```
Though `...\colobot\out\build\x64-Debug\CBot.dll` is produced, there is no `CBot.lib` in the same or any other directory.
CMake command arguments in my case:
```
-Wno-dev -DCMAKE_TOOLCHAIN_FILE=D:\Devel\libs\vcpkg\scripts\buildsystems\vcpkg.cmake -DUSE_STATIC_RUNTIME=1
   -DVCPKG_TARGET_TRIPLET=x64-windows-static -DBOOST_STATIC=1 -DDEV_BUILD=1 -DGLEW_STATIC=1 -DMUSIC=0 
```
`-DMSVC_STATIC=1` seems to be obsolete (isn't used in the scripts) - `USE_STATIC_RUNTIME` is to be used instead.
I thought I had missed something, but according to [this anwer](https://stackoverflow.com/a/71592676/536172) CMake doesn't produce an import library automagically - extra tuning is needed:
```
set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
```
This line should be added to `...\colobot\src\CBot\CMakeLists.txt`:
```
if(CBOT_STATIC)
    add_library(CBot STATIC)
else()
    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
    add_library(CBot SHARED)
    install(TARGETS CBot
            LIBRARY DESTINATION ${COLOBOT_INSTALL_LIB_DIR}
            ARCHIVE DESTINATION ${COLOBOT_INSTALL_LIB_DIR}
            RUNTIME DESTINATION ${COLOBOT_INSTALL_BIN_DIR})
endif()
...
```
After that the import library is generated as `...\colobot\out\build\x64-Debug\src\CBot\CBot.lib`.
Unfortunately the link step isn't successful due to errors with audio libraries:
```
...\out\build\x64-Debug\sndfile.lib(flac.c.obj) :
  error LNK2019: unresolved external symbol FLAC__stream_decoder_new referenced in function flac_read_header
...
...\out\build\x64-Debug\sndfile.lib(ogg.c.obj) :
   error LNK2019: unresolved external symbol ogg_sync_init referenced in function ogg_stream_classify
...
...\out\build\x64-Debug\sndfile.lib(ogg_vorbis.c.obj) :
   error LNK2019: unresolved external symbol vorbis_info_init referenced in function vorbis_read_header
...
...\out\build\x64-Debug\sndfile.lib(ogg_opus.c.obj) :
   error LNK2019: unresolved external symbol opus_get_version_string referenced in function ogg_opus_open
```

**Reply by Anton-V-K** [reply]
The build succeeds only if I link the app with `libsndfile` statically adding `-DSNDFILE_STATIC=1` to CMake command arguments.
Unfortunately the app fails to create `CText` and terminates without entering the main loop.
Here is the log fragment:
```
...
[INFO]: Error on parsing config file: No such node (Experimental.GraphicsDevice)
[INFO]: Creating CDevice - OpenGL 1.4
[INFO]: OpenGL 4.6.0 NVIDIA 466.77
[INFO]: NVIDIA GeForce MX150/PCIe/SSE2
[INFO]: Shadow mapping available (core)
[INFO]: Anisotropic filtering available
[INFO]: Maximum anisotropy: 16
[INFO]: Multisampling supported, max samples: 32
[INFO]: Auto-detecting VBO support
[INFO]: Core VBO supported
[INFO]: Maximum texture units: 4
[INFO]: Maximum texture size: 32768
[DEBUG]: Using quality shadows
[INFO]: Framebuffer supported (ARB)
[INFO]: Maximum renderbuffer size: 32768
[INFO]: CDevice created successfully
[ERROR]: Error creating CText: Error on parsing fonts config file: failed to open file
```
The failure comes from `CFontLoader::Init()` which can't open file `.../fonts/fonts.ini`.
I wonder how the code is supposed to work because the official build doesn't contain such file either. According to the history this file was added to `dev` branch of `data` recently (see  https://github.com/colobot/colobot/commit/73db861cc6e76e1e76797999e154031ed67dab2e).
This issue may also happen on Linux - see #1506 (see also #1031).
To solve it make sure that branches of `colobot` and `data` match each other.

BTW CBot library should also be linked in statically because of static CRT, so `-DCBOT_STATIC=1` should be added to CMake command arguments.
